internal/scanner/walk.go

```diff
@@ -97,6 +101,9 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
 	newParallelHasher(w.Dir, w.BlockSize, w.Hashers, hashedFiles, files)
 
 	go func() {
+		w.symlinksSeen = make(map[string]struct{})
+		w.dirsToScan = nil
```

Comment:
We are sure that no two Walk() operations can happen at the same time, correct?
Seems like a bad smell to let this information change within a go routine. I'd pass this array as a second return value.